Add an AWS EC2 instance fetcher

lxea opened the pull request:
This adds a fetcher for ec2 instances as well as a watcher to periodically fetch them.

Requires IAM permissions:  `iam:DescribeInstances`

Depends on #13787 
Part of #12410

Commit:
Add an AWS EC2 instance fetcher

tigrato commented on lib/srv/db/common/cloud.go:
```diff
@@ -66,6 +68,8 @@ type CloudClients interface {
 	// GetAWSSTSClient returns AWS STS client for the specified region.
 	GetAWSSTSClient(region string) (stsiface.STSAPI, error)
+	// GetAWSEC2Client returns AWS EC2 client for the specified region.
+	GetAWSEC2Client(region string) (ec2iface.EC2API, error)
```
I think we should move it out from `db`, but probably it's better to land it directly in `lib/cloud`.
We will require it for Kubernetes discovery as well.

tigrato commented on lib/srv/regular/sshserver.go:
```diff
@@ -787,6 +790,13 @@ func New(addr utils.NetAddr,
 	}
 
+	if len(s.awsMatchers) != 0 {
+		s.ec2Watcher, err = server.NewCloudServerWatcher(s.awsMatchers)
```
The function call does not match with the signature (missing context and clients)

zmb3 commented on lib/cloud/watchers/ec2.go:
```diff
@@ -0,0 +1,104 @@
+	Document string
+}
+
+func newEc2InstanceFetcher(cfg ec2FetcherConfig) (*ec2InstanceFetcher, error) {
```
Do you need to return an error here? Looks like this can never fail.

zmb3 commented on lib/cloud/watchers/watcher.go:
```diff
@@ -107,16 +133,58 @@ func (w *Watcher) Start() {
 	}
 }
 
+func (w *Watcher) startEC2() {
```
```suggestion
func (w *Watcher) runEC2() {
```

Typically "start" is used for things that kick off some background processing and then return, where "run" is preferred for long running operations that block.

zmb3 reviewed (dismissed):
In addition to the review comments, when you add new code that requires new AWS permissions, please don't forget to document the new permission, update helm/terraform/etc.

greedy52 commented on lib/cloud/watchers/watcher.go:
```diff
@@ -98,6 +119,11 @@ func NewWatcher(ctx context.Context, config WatcherConfig) (*Watcher, error) {
 // subscribe to events such as created/removed instances and tag changes, but
 // this will require Teleport to have more AWS permissions.
 func (w *Watcher) Start() {
+	go w.startEC2()
+	w.start()
+}
```
should we keep the database vs EC2 watcher separate? I assume they will be used for different services.

greedy52 commented on lib/cloud/clients/session.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package common
```
This is a database specific `Session`. I think we should keep it in `lib/srv/db/*`.

r0mant replied:
I agree. I would keep database access related stuff like session, watchers, etc. where they were.

tigrato commented on lib/cloud/watchers/ec2.go:
```diff
@@ -0,0 +1,104 @@
+limitations under the License.
+*/
+
+package watchers
```
I agree with @zmb3 

In my perspective `watchers` should only discover cloud resources and look for updates/deletions. The package should inform the services (can be via callback or channel ping) that a new update exists and provide as well the cloud/type they belong to. The callback/services are responsible for setting and configuring the resource in teleport

r0mant reviewed (changes requested):
I agree with existing comments, I would keep all database access related stuff separate where it was. I think that ec2 should have its own session (if it needs this concept), watcher, etc.

It's ok to move out some common parts to reuse but otherwise I would try not to stuff ec2 stuff into interfaces that were done specifically for database access.

r0mant reviewed (commented):
After discussion with @lxea, here's what we're going to do for this PR:

- Move session and other database related functionality back into srv/db for now so we don't expand the scope of this PR with heavy refactoring. Basically, go back to the original state of the PR which was very compact.
- Move CloudClients into lib/cloud/clients since they can be useful outside of lib/srv/db. Leave their refactoring into per-cloud clients for later.
  - Untangle database session from CloudClients so they can be moved.

In a **separate follow up** PR, in order to not increase the scope of this one, we'll refactor watchers/fetchers to unify them between database and server access in this way:

- Update `Fetcher` interface to return generic types.Resources.
- Update `Watcher` to also use types.Resources when submitting updates to the clients instead of types.Databases.

This will let us implement pluggable watchers that we'll be able to use in database access, server access and for other auto-discovery in future, without having to update existing watchers/fetchers with resource specific "if" cases.

FYI @zmb3 @tigrato

r0mant commented on lib/defaults/defaults.go:
```diff
@@ -767,3 +767,8 @@ const (
 )
+
+const (
+	AWSDefaultInstallerDocument = "TeleportDiscoveryInstaller"
```
```suggestion
        // AWSInstallerDocument is ...
	AWSInstallerDocument = "TeleportDiscoveryInstaller"
```

r0mant commented on lib/defaults/defaults.go:
```diff
@@ -767,3 +767,8 @@ const (
+
+const (
+	AWSDefaultInstallerDocument = "TeleportDiscoveryInstaller"
+	DefaultIamInviteTokenName   = "aws-discovery-iam-token"
```
```suggestion
        // IAMInviteTokenName is ...
	IAMInviteTokenName   = "aws-discovery-iam-token"
```

r0mant commented on lib/services/matchers.go:
```diff
@@ -37,9 +44,13 @@ type AWSMatcher struct {
+	JoinToken string
+	// SSM provides options to use when sending a document command to
+	// an EC2 node
+	SSM *AWSSSM `yaml:"ssm,omitempty"`
```
Does this field need a json tag? Others don't have it (I don't think this struct is marshaled/unmarshaled).

r0mant commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/cloudflare/cfssl/log"
```
I don't think you meant to import this log package.

r0mant commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+	"github.com/gravitational/trace"
+)
+
+type EC2Instances struct {
```
Add godocs on all public names/fields in this file.

r0mant commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+type Watcher struct {
+	// Instances can be used to consume
```
Unfinished comment.

r0mant commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+
+type Watcher struct {
+	// Instances can be used to consume
+	Instances chan EC2Instances
```
Nit: `InstancesC` would be a more customary name for this field since it's a channel.
```suggestion
	InstancesC chan EC2Instances
```

r0mant commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+func (w *Watcher) Run() {
+	ticker := time.NewTicker(w.waitTime)
```
```suggestion
	ticker := time.NewTicker(w.waitTime)
	defer ticker.Close()
```

r0mant commented on lib/srv/regular/sshserver.go:
```diff
@@ -214,8 +216,14 @@ type Server struct {
 	// awsMatchers are used to match EC2 instances
 	awsMatchers []services.AWSMatcher
+	// cloudClients are clients used for fetching cloud resources
+	cloudClients cloud.Clients
```
Why does the `Server` need cloud clients as a field?

r0mant commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+				log.Error("Failed to fetch EC2 instances: ", err)
+				continue
+			}
+			w.Instances <- *inst
```
Should select on `w.ctx.Done()` as well.

r0mant commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+	Instances chan EC2Instances
+
+	fetchers []*ec2InstanceFetcher
+	waitTime time.Duration
```
Nit: `waitTime` is a little non-descriptive, I'd maybe call it `fetchInterval`.

r0mant commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+	Document string
+}
+
+func newEc2InstanceFetcher(matcher services.AWSMatcher, region, document string,
```
```suggestion
func newEC2InstanceFetcher(matcher services.AWSMatcher, region, document string,
```

r0mant commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+	w.cancel()
+}
+
+func NewCloudServerWatcher(ctx context.Context, matchers []services.AWSMatcher, clients cloud.Clients) (*Watcher, error) {
```
@lxea I don't see where this watcher is actually being used. Is it in a different PR?

r0mant reviewed (commented):
@tigrato @greedy52 Could you guys take another look here as well please?

greedy52 commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+	ec2Client ec2iface.EC2API, labels types.Labels) *ec2InstanceFetcher {
+	tagFilters := make([]*ec2.Filter, 0, len(labels)+1)
+	tagFilters = append(tagFilters, &ec2.Filter{
+		Name:   aws.String("instance-state-name"),
```
nitpick: could you make a constant for `instance-state-name` with a description where it comes from? thanks

greedy52 commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+			Name:   aws.String("tag:" + key),
+			Values: aws.StringSlice(val),
+		})
+	}
```
This probably wasn't in the original scope but I suggest having an optional **`vpc-id`** filter (through a hard-coded label, or a separate field like `same_vpc: true`). I think it can be quite convenient in combination with the resource tags (especially when EC2 resource tags are lacking).

What do you think? @lxea @r0mant @tigrato

tigrato replied:
I agree with you @greedy52 
This will allow them to filter out instances that otherwise can be there like EKS-managed nodes...

r0mant replied:
@greedy52 @tigrato Instead of introducing an extra field, can't the same be accomplished with using node tags?

r0mant replied:
Let's push it out of scope for now.

greedy52 commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,150 @@
+		for _, fetcher := range w.fetchers {
+			inst, err := fetcher.GetEC2Instances(w.ctx)
+			if err != nil {
+				log.Error("Failed to fetch EC2 instances: ", err)
```
When the agent has no permission to fetch, this log will spam every time it tries. In the db cloud fetcher, we have downgraded `AccessDenied` errors to a debug log.

tigrato commented on lib/config/fileconf.go:
```diff
@@ -437,6 +437,37 @@ func (conf *FileConfig) CheckAndSetDefaults() error {
+		}
+
+		if matcher.SSM.Document == "" {
+			matcher.SSM.Document = defaults.AWSInstallerDocument
```
doesn't SSM Doc need to be manually created for each AWS account? 
If that's the case should we force default values or should we just return an error either if empty or if the SSM doc does not exist

Same logic applies for IAM Token Name

tigrato commented on lib/srv/regular/sshserver.go:
```diff
@@ -805,6 +811,13 @@ func New(addr utils.NetAddr,
 	}
 
+	if len(s.awsMatchers) != 0 {
+		s.cloudWatcher, err = server.NewCloudServerWatcher(s.ctx, s.awsMatchers, cloud.NewClients())
```
For testing purposes shouldn't we receive the cloud clients as`ServerOption`  and use the `cloud.NewClients()` function to configure the cloudClients if not specified in option args received?

tigrato commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,156 @@
+		for _, fetcher := range w.fetchers {
+			inst, err := fetcher.GetEC2Instances(w.ctx)
+			if err != nil {
+				log.Error("Failed to fetch EC2 instances: ", err)
```
use `WithError(err).Error(...)` so the stack trace can be correctly printed

tigrato commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,156 @@
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			fetcher :=
```
same line

tigrato commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,156 @@
+			}
+			fetcher :=
+				newEC2InstanceFetcher(matcher, region, matcher.SSM.Document, cl, matcher.Tags)
+			if err != nil {
```
This error check is not required, at least with the current config

tigrato commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,156 @@
+		ctx:           cancelCtx,
+		cancel:        cancelFn,
+		fetchInterval: time.Minute,
+		InstancesC:    make(chan EC2Instances),
```
maybe create a buffered channel otherwise both producers and consumers have to sync for each discovered ec2 instance

tigrato commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,156 @@
+
+func newEC2InstanceFetcher(matcher services.AWSMatcher, region, document string,
+	ec2Client ec2iface.EC2API, labels types.Labels) *ec2InstanceFetcher {
+	tagFilters := make([]*ec2.Filter, 0, len(labels)+1)
```
if you know exactly the size of the array you don't need to rely on append

r0mant commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,197 @@
+func (f *ec2InstanceFetcher) GetEC2Instances(ctx context.Context) (*EC2Instances, error) {
+	var instances []*ec2.Instance
+	var accountID string
+	err := f.EC2.DescribeInstancesPagesWithContext(ctx, &ec2.DescribeInstancesInput{
```
This fetches all EC2 nodes matching the filters on every "tick" of the watcher, correct? I wonder if it's going to cause issues with large fleets of nodes.

We can probably address this out of scope of this particular PR, but I wonder if we can mark every AWS instance that we've already enrolled with some tag (e.g. `TeleportAutoEnrolled: true`) and then filter them out during this discovery. What do you think?

r0mant reviewed (approved):
Looks good to me, the main remaining comment I have is about excluding nodes that already have been enrolled from the discovery but that is more of an optimization and can be addressed in a follow up.

@greedy52 @tigrato Do you guys have any other feedback on this one?

Commit:
Resolve comments

- Remove leftover VPCID stuff
- Document public functions
- Return a notfound when there are no nodes fetched instead of nil nil

greedy52 commented on lib/srv/server/cloudwatcher.go:
```diff
@@ -0,0 +1,195 @@
+		func(dio *ec2.DescribeInstancesOutput, b bool) bool {
+			for _, res := range dio.Reservations {
+				if accountID == "" {
+					accountID = aws.StringValue(res.OwnerId)
```
What is the use of `AccountID`?  If my understanding is correct, the `OwnerId` is the owner of the reservation.

Commit:
Resolve comments

Commit:
Fix fileconf_test.go
